File: 0599-minimum-index-sum-of-two-lists/0599-minimum-index-sum-of-two-lists.py
class Solution:
    def findRestaurant(self, list1: List[str], list2: List[str]) -> List[str]:
        mapping = {}
        ans = 3000
        list = []
        for i in range(len(list1)):
            mapping[list1[i]] = i
        for j in range(len(list2)):
            if list2[j] in mapping:
                if mapping[list2[j]]+j < ans:
                    ans = j + mapping[list2[j]]
                    list = []
                    list.append(list2[j])
                elif mapping[list2[j]]+j == ans:
                    list.append(list2[j])

        return list

File: 0599-minimum-index-sum-of-two-lists/0599-minimum-index-sum-of-two-lists.java
class Solution {
    public String[] findRestaurant(String[] list1, String[] list2) {
        Map<String,Integer> map = new HashMap<>();
        List<String> list = new ArrayList<>();
        int ans = 3000;
        for(int i=0;i<list1.length;i++)
            map.put(list1[i],i);
        for(int i=0;i<list2.length;i++){
            if (!map.containsKey(list2[i]))
                continue;
            if(i+map.get(list2[i]) < ans){
                ans = i+map.get(list2[i]);
                list.clear();
                list.add(list2[i]);
            }else if(i+map.get(list2[i]) == ans)
                list.add(list2[i]);
        }       
        return list.toArray(new String[0]);
    }
}